File: report_exercises/ex4_binary_vision.py
import numpy as np
import cv2

img = cv2.imread('rice.png')
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
_, otsu = cv2.threshold(gray, 200, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)

kernel = np.ones((6, 6), np.uint8)
eroded = cv2.erode(otsu, kernel)

analysis = cv2.connectedComponentsWithStats(eroded, 10, cv2.CV_32S)
labels, labelIds, values, centroid = analysis
print(labels)

cv2.imshow('', cv2.resize(eroded, (0, 0), fx=2, fy=2))
cv2.waitKey(0)

File: report_exercises/ex4_binary_vision.cpp
#include <stdio.h>
#include <opencv2/opencv.hpp>

using namespace std;

int countNumberOfRice(cv::Mat input) {
  cv::Mat binary, eroded, labels, stats;
  cv::threshold(input, binary, 150, 255, 0);
  cv::erode(binary, eroded, 2);
  // connectedComponents don't work!!
  // int i, nComps = cv::connectedComponentsWithStats(eroded, labels, stats, cv::noArray());

  cv::namedWindow("Image");
  cv::imshow("Image", eroded);
  cv::waitKey(0);
  return 0;
}

int main(int argc, char* argv[]) {
  cv::Mat img = cv::imread("D:/machine-vision/rice.png", cv::IMREAD_GRAYSCALE);
  if (!img.data) {
    printf("No image \n");
    return -1;
  }
  int numOfRice = countNumberOfRice(img);
  cout << numOfRice << endl;
  return 0;
}
